src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,478 @@
+            // This branch should be unreachable because we create the file if it doesn't exist
+            printMessage(MESSAGE_DATA_FILE_ERROR);
+            System.exit(-2);
+        }
+    }
+
+    /**
+     * Parse task saved as string format and convert it back into an object.
+     * @param taskAsString task saved as string
+     * @param lineNum line number
+     * @return Task object if successful, or null if error
+     */
+    private static Task parseSavedTask(String taskAsString, int lineNum) {
+        String[] processedString = taskAsString.split(FS);
+        try {
+            final String taskType = processedString[0];
+            final boolean taskMarked;
+            if (processedString[1].equals("1")) {
+                taskMarked = true;
+            } else if (processedString[1].equals("0")) {
+                taskMarked = false;
+            } else {
+                throw new MalformedTaskFormatException();
+            }
+            final String taskDescription = processedString[2];
+            switch (taskType) {
+            case "T":
+                if (processedString.length != 3) {
+                    throw new MalformedTaskFormatException();
+                }
+                Todo todo = new Todo(taskDescription);
+                todo.setIsDone(taskMarked);
+                return todo;
+            case "D":
+                if (processedString.length != 4) {
+                    throw new MalformedTaskFormatException();
+                }
+                final String dueBy = processedString[3];
+                Deadline deadline = new Deadline(taskDescription, dueBy);
+                deadline.setIsDone(taskMarked);
+                return deadline;
+            case "E":
+                if (processedString.length != 4) {
+                    throw new MalformedTaskFormatException();
+                }
+                final String time = processedString[3];
+                Event event = new Event(taskDescription, time);
+                event.setIsDone(taskMarked);
+                return event;
+            default:
+                throw new MalformedTaskFormatException();
+            }
+        } catch (IndexOutOfBoundsException | MalformedTaskFormatException e) {
+            // Using System.out.println here because I don't want to print a horizontal line after each output
+            System.out.println(String.format(MESSAGE_MALFORMED_TASK, lineNum, taskAsString));
+            return null;
+        }
+    }
+
+    /**
+     * Attempt to save current task data to data file.
+     */
+    private static void saveToFile() {
+        try {
+            FileWriter fw = new FileWriter(dataFile);
+            for (Task t : savedTasks) {
+                fw.write(t.formatAsData(FS) + LS);
+            }
+            fw.close();
+        } catch (IOException e) {
+            // Using System.out.println here because I don't want to print a horizontal line after each output
```

Comment:
Might want to remove this comment since the implication of using `System.out.println` should be straightforward.